buy_together_app/tests/test_sign_up.py

```diff
@@ -0,0 +1,72 @@
+import pytest
+from django.template.loader import get_template
+from supplier.models import Supplier
+from django.contrib.auth.models import User
+
+
+@pytest.mark.django_db()
+class TestSignupPage:
+    def test_signup_template(self):
+        get_template('buy_together_app/sign_up.html')
+
+    def test_signup_supplier(self, client):
+        response = client.post(
+            '/signup/supplier', {'username': "testSignUp",
+                                 'password': "pokemon123",
+                                 'first_name': "ash",
+                                 'last_name': "katchamp",
+                                 'business_name': "pokemon inc"}
+        )
+        assert User.objects.get(username="testSignUp") in list(User.objects.all())
+        assert Supplier.objects.get(
+            supplier_account=User.objects.get(username="testSignUp")).business_name == "pokemon inc"
+        assert response.status_code == 302
+        assert response.url == "/"
+
+    def test_bad_signup_supplier(self, client, saved_supplier0):
+        response = client.post(
+            '/signup/supplier', {'username': saved_supplier0.supplier_account.username,
+                                 'password': saved_supplier0.supplier_account.password,
+                                 'first_name': saved_supplier0.supplier_account.first_name,
+                                 'last_name': saved_supplier0.supplier_account.last_name,
+                                 'business_name': saved_supplier0.business_name}
+        )
+        # In here we are checking that form is valid but the supplier is
+        # already saved in the data base
+        assert "Username is taken" in response.context['message']
+
+    def test_signup_client(self, client):
+        response = client.post(
+            '/signup/client', {'username': "testSignUp",
+                               'password': "pokemon123",
+                               'first_name': "ash",
+                               'last_name': "katchamp",
+                               'area': "pokemon inc"}
+        )
+        assert User.objects.get(username="testSignUp") in list(User.objects.all())
+        assert response.status_code == 302
+        assert response.url == "/"
+
+    def test_bad_signup_client(self, client, saved_client0):
+        response = client.post(
+            '/signup/client', {'username': saved_client0.client_account.username,
+                               'password': saved_client0.client_account.password,
+                               'first_name': saved_client0.client_account.first_name,
+                               'last_name': saved_client0.client_account.last_name,
+                               'area': saved_client0.area}
+        )
+        # In here we are checking that form is valid but the client is
+        # already saved in the data base
+        assert "Username is taken" in response.context['message']
+
+    def test_signup_invalid(self, client):
```

Comment:
What makes this invalid? The fact that username is not provided or the fact that you're providing a field that does not exist? If it's the missing input, then better to remove the other input, if it's an unknown input, I suggest adding the missing input and giving this one a very different name (e.g. `unknown_field`), if both scenarios should fail, I suggest creating multiple test cases.

Reply:
In the front you can not send a request untill the form is valid which is a valid post request but this situation if you use a post request not from the front of the page but using something like PostMan so you can post request using invalid fields and if that is the case i handle it

Reply:
I understand that, but it needs to be clear what makes this call invalid.